packages/core/contracts/merkle-distributor/implementation/MerkleDistributor.sol

```diff
@@ -0,0 +1,236 @@
+// SPDX-License-Identifier: AGPL-3.0-only
+
+/**
+ * Inspired by:
+ * - https://github.com/pie-dao/vested-token-migration-app
+ * - https://github.com/Uniswap/merkle-distributor
+ * - https://github.com/balancer-labs/erc20-redeemable
+ *
+ * @title MerkleDistributor contract.
+ * @notice Allows an owner to distribute any reward ERC20 to claimants according to Merkle roots. The owner can specify
+ *         multiple Merkle roots distributions, each of which has its own start time constraining when claims can be
+ *         executed.
+ */
+
+pragma solidity ^0.6.0;
+pragma experimental ABIEncoderV2;
+
+import "@openzeppelin/contracts/token/ERC20/SafeERC20.sol";
+import "@openzeppelin/contracts/cryptography/MerkleProof.sol";
+import "@openzeppelin/contracts/access/Ownable.sol";
+import "@openzeppelin/contracts/token/ERC20/IERC20.sol";
+import "../../common/implementation/Lockable.sol";
+import "../../common/implementation/Testable.sol";
+
+contract MerkleDistributor is Ownable, Lockable, Testable {
+    using SafeMath for uint256;
+    using SafeERC20 for IERC20;
+
+    // A Window maps a Merkle root to a reward token address and a claim window.
+    // Claims for this window cannot take place before start time.
+    struct Window {
+        // Claims for this window can begin after `start`.
+        uint256 start;
+        // Merkle root describing the distribution.
+        bytes32 merkleRoot;
+        // Currency in which reward is processed.
+        IERC20 rewardToken;
+        // Total amount of rewards distributed this window. This is not enforced
+        // but might be useful to query.
+        uint256 totalRewardsDistributed;
```

Comment:
nit: thoughts on changing this to just `rewards`, or `totalRewards`?

Would also like to double check that this variable isn't causing any increase in gas costs during the claim. I only ask because I could see solidity doing something dumb like reading this slot when the struct is accessed even though _this slot_ isn't explicitly used in the code. If that is the case, I would argue that we should rely on the event to access this var. Otherwise, definitely on board with keeping it.

Reply:
Yeah this variable is mostly just a sanity check, or a convenience variable, so I'm happy to remove it.

Reply:
I removed this variable because there is no strict reason why it needs to be stored on chain. It could easily be stored offchain on IPFS for example

Reply:
SGTM. I've got this in the SDK right now already.